runelite-client/src/main/java/net/runelite/client/plugins/chatfilter/ChatFilterPlugin.java

```diff
@@ -0,0 +1,209 @@
+import java.util.List;
+import java.util.Objects;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+import javax.inject.Inject;
+import net.runelite.api.ChatMessageType;
+import net.runelite.api.Client;
+import net.runelite.api.Player;
+import net.runelite.api.events.ConfigChanged;
+import net.runelite.api.events.OverheadTextChanged;
+import net.runelite.api.events.ScriptCallbackEvent;
+import net.runelite.client.config.ConfigManager;
+import net.runelite.client.eventbus.Subscribe;
+import net.runelite.client.plugins.Plugin;
+import net.runelite.client.plugins.PluginDescriptor;
+import net.runelite.client.util.Text;
+import org.apache.commons.lang3.StringUtils;
+
+@PluginDescriptor(
+	name = "Chat Filter",
+	description = "Censor user configurable words or patterns from chat",
+	enabledByDefault = false
+)
+public class ChatFilterPlugin extends Plugin
+{
+	private static final Splitter NEWLINE_SPLITTER = Splitter
+		.on("\n")
+		.omitEmptyStrings()
+		.trimResults();
+
+	private static final String CENSOR_MESSAGE = "Hey, everyone, I just tried to say something very silly!";
+
+	private final List<Pattern> filteredPatterns = new ArrayList<>();
+
+	@Inject
+	private Client client;
+
+	@Inject
+	private ChatFilterConfig config;
+
+	@Provides
+	ChatFilterConfig provideConfig(ConfigManager configManager)
+	{
+		return configManager.getConfig(ChatFilterConfig.class);
+	}
+
+	@Override
+	protected void startUp() throws Exception
+	{
+		updateFilteredPatterns();
+	}
+
+	@Override
+	protected void shutDown() throws Exception
+	{
+		filteredPatterns.clear();
+	}
+
+	@Subscribe
+	public void onScriptCallbackEvent(ScriptCallbackEvent event)
+	{
+		if (!"chatFilterCheck".equals(event.getEventName()))
+		{
+			return;
+		}
+
+		int[] intStack = client.getIntStack();
+		int intStackSize = client.getIntStackSize();
+		ChatMessageType chatMessageType = ChatMessageType.of(intStack[intStackSize - 1]);
+
+		// Only filter public chat and private messages
+		switch (chatMessageType)
+		{
+			case PUBLIC:
+			case PUBLIC_MOD:
+			case AUTOCHAT:
+			case PRIVATE_MESSAGE_RECEIVED:
+			case PRIVATE_MESSAGE_RECEIVED_MOD:
+			case CLANCHAT:
+				break;
+			default:
+				return;
+		}
+
+		String[] stringStack = client.getStringStack();
+		int stringStackSize = client.getStringStackSize();
+
+		String message = stringStack[stringStackSize - 1];
+		String censoredMessage = censorMessage(message);
+
+		if (censoredMessage == null)
+		{
+			// Block the message
+			intStack[intStackSize - 2] = 0;
+		}
+		else
+		{
+			// Replace the message
+			stringStack[stringStackSize - 1] = censoredMessage;
+		}
+	}
+
+	@Subscribe
+	public void onOverheadTextChanged(OverheadTextChanged event)
+	{
+		if (!(event.getActor() instanceof Player))
+		{
+			return;
+		}
+
+		String message = censorMessage(event.getOverheadText());
+
+		event.getActor().setOverheadText(message);
+	}
+
+	String censorMessage(String message)
```

Comment:
as per discussion in Discord this should filter invisible symbols, otherwise this whole thing will not work half of the time as spammers use them often

Reply:
Do we have a way to know which characters are non printable?